### ISSUE TITLE
Possible alignment issue?

### ISSUE THREAD

**Issue by awnumar** [prologic/bitcask#61]
Running the benchmarks on my system, I get unpredictable drops in `Put` performance for some runs.

```
BenchmarkPut/128B-2         	  500000	      3285 ns/op	  38.96 MB/s
BenchmarkPut/256B-2         	  300000	      4178 ns/op	  61.26 MB/s
BenchmarkPut/1K-2           	  200000	      9387 ns/op	 109.08 MB/s
BenchmarkPut/2K-2           	  100000	     29537 ns/op	  69.34 MB/s
BenchmarkPut/4K-2           	   50000	     30926 ns/op	 132.44 MB/s
BenchmarkPut/8K-2           	   20000	     61055 ns/op	 134.17 MB/s
BenchmarkPut/16K-2          	   10000	    117457 ns/op	 139.49 MB/s
BenchmarkPut/32K-2          	   10000	    356458 ns/op	  91.93 MB/s
```
```
BenchmarkPut/128B-2         	  500000	      3311 ns/op	  38.66 MB/s
BenchmarkPut/256B-2         	  300000	      4481 ns/op	  57.13 MB/s
BenchmarkPut/1K-2           	  200000	      9326 ns/op	 109.80 MB/s
BenchmarkPut/2K-2           	  100000	     16315 ns/op	 125.52 MB/s
BenchmarkPut/4K-2           	   50000	     58090 ns/op	  70.51 MB/s
BenchmarkPut/8K-2           	   30000	     59509 ns/op	 137.66 MB/s
BenchmarkPut/16K-2          	   10000	    112701 ns/op	 145.37 MB/s
BenchmarkPut/32K-2          	   10000	    221304 ns/op	 148.07 MB/s
```

**Comment by prologic** [comment]
> Possible alignment issue?

What do you mean by "alignment'"?

**Comment by awnumar** [comment]
Since the spikes happen sporadically for specific tests, I suspect it could be due to page-alignment issues or some related memory issue that results in a missed optimisation by the compiler. I could be completely off the mark with that though.

**Comment by prologic** [comment]
Yeah you're talking about CPU cache hits/misses I think? To be honest I'm pretty rought when it comes to that level of low/deep optimization. I've read wonderful blog posts on the subject though; it is worth investigating for sure!

**Comment by awnumar** [comment]
There's multiple cache levels so it could be happening in virtual memory as well.

**Comment by prologic** [comment]
> It would be helpful to know if you can reproduce this?

Yeah so I can repro inconsistent results here too. The question I have is this:

> Are we really that concerned with ~100-200ns  fluctuations?

Don't get me wrong I _think_ its still worth investigating further; I just have no active experience in this level of optimization :) Need to go do some digging!

**Comment by prologic** [comment]
Also what tools  are available to us to help investigate this?

**Comment by awnumar** [comment]
> Are we really that concerned with ~100-200ns fluctuations?

A goal of bitcask is consistent and predictable performance, so I would say so yes. At least figuring out why.

> Also what tools are available to us to help investigate this?

I'm not sure. Go's `pprof` might do the trick, going to look into it later.

> I just have no active experience in this level of optimization

Me neither, but as I said, I could be completely wrong about the cause. The issue seems to have started with the latest changes so that's a good starting point.

**Comment by awnumar** [comment]
Lol reproduced on my phone

![image](https://user-images.githubusercontent.com/8942495/62697802-827d5280-b9d3-11e9-9975-040b96d6ae6b.png)

**Comment by prologic** [comment]
This [post](https://medium.com/@felipedutratine/how-to-organize-the-go-struct-in-order-to-save-memory-c78afcf59ec20 is worth a read on the subject.

**Comment by awnumar** [comment]
Ah classic padding issue. Unfortunately, running [`maligned`](https://github.com/mdempsky/maligned) returned nothing.

**Comment by prologic** [comment]
So I can _gain_ some small improvement by reordering the fields of `Item` and `Entry` and adding some padding. Here are y results:

Before:
```
>>> xs = [3814,3594,3767,3495,3495,3807,3907,3773,4104,4222]
>>> mean(xs)
3797.8
>>> median(xs)
3790.0
>>> stdev(xs)
238.6186171371472
```

After:
```
>>> xs = [3713, 3687, 3526, 3377, 3524, 3543, 3488, 3862, 4129, 3590]
>>> mean(xs)
3643.9
>>> median(xs)
3566.5
>>> stdev(xs)
217.67072482179233
```

I guess this a ~8-10% improvement? But I'm not convinced given `maligned` turns up empty.

**Comment by awnumar** [comment]
Looking at the cpu trace, [prof.zip](https://github.com/prologic/bitcask/files/3481428/prof.zip), it could perhaps be `(*bufio.Writer).Flush()` that is to blame.

If this is true, perhaps a queue system using a buffered channel with a receiving goroutine handling the actual writes to disk could prevent the occasional sync block.

> But I'm not convinced given maligned turns up empty.

Perhaps maligned missed it.

**Comment by prologic** [comment]
Actually you might be onto something there; there seems to be very little to be gained from reordering fields in the two structs in constant use `Item` and `Entry` -- I mean I see a small improvement but its more than likely the OS flushing buffers and causing blocking here.

Can we clarify that you were referring to the inconsistent performance of the `.Put()` call?

**Comment by prologic** [comment]
> If this is true, maybe a queue system using a buffered channel with a receiving goroutine handling the actual writes to disk could prevent the occasional sync block.
> 
> >

I would support a PR that implements this anyway; I've always wanted to allow the user of Bitcask a few durability / performance tradeoffs/options:

- Default -- Let the OS flush buffers when it wants to (_what we have now_)
- Strong -- Always Sync/Flush buffers to disk (_will be slower writes_)
- Queue/Interval -- Perform Sync/Flush in the background (_what you described_)

**Comment by awnumar** [comment]
> Can we clarify that you were referring to the inconsistent performance of the .Put() call?

Yes.

**Comment by prologic** [comment]
😱Adding a `WithSync(true)` option forcing a `.Sync()` on the active data file (_the write-ahead log_) slows down write performance to a whopping ~200/s crawl!

Also the consistency off the charts!

```#!python
>>> from statistics import mean, median, stdev
>>> xs = [4421023,4636717,4750086,4686166,4736949,4127772,4593664,4561493,4580352,4636388]
>>> len(xs)
10
>>> mean(xs)
4573061.0
>>> median(xs)
4615026.0
>>> stdev(xs)
182896.92644826544
>>> 1e9 / mean(xs)
218.6719136263435
>>>
```

I'm running the tests like this:

```#!sh
$ for i in $(seq 0 9); do go test -run=BenchmarkPut/128B -benchmem -bench=BenchmarkPut/128B .; done | grep -o -E  '\d+ ns\/op' | cut -f 1 -d ' ' | paste -s -d,
4421023,4636717,4750086,4686166,4736949,4127772,4593664,4561493,4580352,4636388
```

The diff looks like this:

```#!diff
diff --git a/bitcask.go b/bitcask.go
index 3707ed6..f21baf2 100644
--- a/bitcask.go
+++ b/bitcask.go
@@ -156,6 +156,12 @@ func (b *Bitcask) Put(key, value []byte) error {
 		return err
 	}
 
+	if b.config.sync {
+		if err := b.curr.Sync(); err != nil {
+			return err
+		}
+	}
+
 	item := b.keydir.Add(key, b.curr.FileID(), offset, n)
 	b.trie.Add(string(key), item)
 
diff --git a/bitcask_test.go b/bitcask_test.go
index df79488..edef130 100644
--- a/bitcask_test.go
+++ b/bitcask_test.go
@@ -572,7 +572,7 @@ func BenchmarkPut(b *testing.B) {
 	}
 	defer os.RemoveAll(testdir)
 
-	db, err := Open(testdir)
+	db, err := Open(testdir, WithSync(true))
 	if err != nil {
 		b.Fatal(err)
 	}
diff --git a/options.go b/options.go
index 74dc2a4..64c17f3 100644
--- a/options.go
+++ b/options.go
@@ -24,25 +24,29 @@ type config struct {
 	maxDatafileSize int
 	maxKeySize      int
 	maxValueSize    int
+	sync            bool
 }
 
 func (c *config) MarshalJSON() ([]byte, error) {
 	return json.Marshal(struct {
-		MaxDatafileSize int `json:"max_datafile_size"`
-		MaxKeySize      int `json:"max_key_size"`
-		MaxValueSize    int `json:"max_value_size"`
+		MaxDatafileSize int  `json:"max_datafile_size"`
+		MaxKeySize      int  `json:"max_key_size"`
+		MaxValueSize    int  `json:"max_value_size"`
+		Sync            bool `json:"sync"`
 	}{
 		MaxDatafileSize: c.maxDatafileSize,
 		MaxKeySize:      c.maxKeySize,
 		MaxValueSize:    c.maxValueSize,
+		Sync:            c.sync,
 	})
 }
 
 func getConfig(path string) (*config, error) {
 	type Config struct {
-		MaxDatafileSize int `json:"max_datafile_size"`
-		MaxKeySize      int `json:"max_key_size"`
-		MaxValueSize    int `json:"max_value_size"`
+		MaxDatafileSize int  `json:"max_datafile_size"`
+		MaxKeySize      int  `json:"max_key_size"`
+		MaxValueSize    int  `json:"max_value_size"`
+		Sync            bool `json:"sync"`
 	}
 
 	var cfg Config
@@ -60,6 +64,7 @@ func getConfig(path string) (*config, error) {
 		maxDatafileSize: cfg.MaxDatafileSize,
 		maxKeySize:      cfg.MaxKeySize,
 		maxValueSize:    cfg.MaxValueSize,
+		sync:            cfg.Sync,
 	}, nil
 }
 
@@ -94,3 +99,12 @@ func WithMaxValueSize(size int) Option {
 		return nil
 	}
 }
+
+// WithSync causes Sync() to be called on every key/value written increasing
+// durability and saftey at the expense of performance
+func WithSync(sync bool) Option {
+	return func(cfg *config) error {
+		cfg.sync = sync
+		return nil
+	}
+}
```

**Comment by awnumar** [comment]
What I'm wondering is why this issue didn't crop up before the latest set of changes.

**Comment by prologic** [comment]
Too focused on read performance? :)

**Comment by prologic** [comment]
Q: Should adding a `.Sync()` after every `.Write()` cause a ~400x decrease in performance? Are SSDs not as great as people make them out to be? :)

**Comment by awnumar** [comment]
> Q: Should adding a .Sync() after every .Write() cause a ~400x decrease in performance? Are SSDs not as great as people make them out to be? :)

That is about what I would expect. SSDs are fast but DRAM is way faster. Those buffered writes go to memory and the sync has to "stop the world" to write the backlog to disk before continuing. This is why a non-blocking queue might work well.

**Comment by prologic** [comment]
Okay well I have the diff for `WithSync()` I can put up a PR for that. Do you want to build the non-blocking queue thingie and see if that helps?

**Comment by awnumar** [comment]
@prologic Sure.

**Comment by jsign** [comment]
I also confirm seeing that drops. I don't have a confirming diagnostic of why this happens, but I think the following things are in my mind:
* The `Put` benchmark always uses the same `db` instance with the `DefaultMaxDatafileSize`. This means that depending on `b.N`, and also in the sub-case test, it will create multiple files in the course of the benchmark. This means that the benchmark is not very predictable in the number of file creations that it does.
* The `Encoder` uses a `bufio.NewWriter` which is the same as `bufio.NewWriterSize` with the `bufio` default size for the buffer (4096). In subtests which imply large writes (such as 32K), that buffer size is too small. This means that the `Write(msg.Value)` will always exceed the buf size, and AFAIK this means that it will directly write to the underlying `Writer` (basically, an indirect flush).

Unfortunately, I tested some ideas regarding these things:
* Using a big `bufio.NewWriterSize`
* Commenting the code that checks if the file size is greater than the maximum allowed to see if the creation of the rolling file was the problem.
* Instead of doing multiple `Write` in the encoder, join all `[]byte` in a single slice and do a single `Write`.

None of these ideas (or even some combinations of them) avoided the throughput drops.

I checked a commit from 10 days ago and seems that this problem already existed:
```
(9123716)⚡ % go test -v -run none -bench=BenchmarkPut/32K -benchtime=2s                                                                     ~/code/bitcask
goos: linux
goarch: amd64
pkg: github.com/prologic/bitcask
BenchmarkPut/32K-8                 10000            290570 ns/op         112.77 MB/s
PASS
ok      github.com/prologic/bitcask     4.534s
(9123716)⚡ % go test -v -run none -bench=BenchmarkPut/32K -benchtime=2s                                                                     ~/code/bitcask
goos: linux
goarch: amd64
pkg: github.com/prologic/bitcask
BenchmarkPut/32K-8                 10000           1000192 ns/op          32.76 MB/s
```

Somehow I think that the unpredictability of the `Flush` using the `bufio.Writer` with the default bufsize value is responsible for this (not only the explicit `Flush` that is done at the end, but also the implicit `Flush` in the `Write` implementation when `len(data)>cap(buf)` (see [here](https://github.com/golang/go/blob/release-branch.go1.12/src/bufio/bufio.go#L617)).

Not definite conclusion, just brainstorming.

**Comment by prologic** [comment]
For better/easier testing:

```#!sh
$ xs="$(for i in $(seq 0 9); do go test -run=BenchmarkPut/128B -bench=BenchmarkPut/128B .; done | grep -o -E  '\d+ ns\/op' | cut -f 1 -d ' ' | paste -s -d,)"; python -c "from statistics import mean, stdev; xs = [$xs]; print mean(xs), stdev(xs)"
11838.3 386.446503413
```

**Comment by prologic** [comment]
Hmmm Am I doing something wrong? I see that `.Get()` for 128B values shows the same deviations?

```#!sh
$ xs="$(for i in $(seq 0 9); do go test -run=BenchmarkGet/128B -bench=BenchmarkGet/128B .; done | grep -o -E  '\d+ ns\/op' | cut -f 1 -d ' ' | paste -s -d,)"; python -c "from statistics import mean, stdev; xs = [$xs]; print mean(xs), stdev(xs)"
3752.8 311.633616786
```

**Comment by jsign** [comment]
In my case `.Get() 128` is very stable.

**Comment by prologic** [comment]
I do wonder how much of this write performance variation comes from the HW + OS itself.

**Comment by prologic** [comment]
I'm not sure what else we can do here? The issue started out with a claim of "memory alignment" but so far we agree its more to do with some perhaps expected system behaviour with flushing/syncing data to disk.

So my question now is... Now what? I'm all out of ideas :)

**Comment by jsign** [comment]
Same here for the moment. I'll try to find someone some other gophers that might take a look.

**Comment by prologic** [comment]
That would be good if we can get some other interested Gopher to have a look at this :)

**Comment by jsign** [comment]
Well, some smart guys rapidly identified the biggest problem as `fsync`ing so many times. That introduces a lot of volatility in the times since `fsync` not only synces writes of the current process, but also it may sync pending writes of other processes too. Also, it seems that the exact guarantee that `fsync` gives is quite weak and it may even lie. I was referred to an interesting [article](https://danluu.com/deconstruct-files/) that I didn't have the chance to read yet.

I thought of mocking the `io.Writer` to bound possible latencies there since the idea of this benchmark isn't to test the performance of disk writes, so seems fair. But one gopher gave simpler advice: use `/tmp` to store the database files and mount `/tmp` in RAM (which I did editing `/etc/fstab`). 

After mounting `/tmp` in RAM, and only changing `bitcask_test.go` to store the files in `/tmp` instead of `pwd`, the results seem to be very predictable.

In short, the unpredictability comes from syncs to disk... which we shouldn't be so worried about since doesn't depend on the code. But I subscribe to the idea that you discussed earlier of maybe capping `fsync` since it may hurt further.

**Comment by jsign** [comment]
I found two interesting related discussions: [1](https://github.com/golang/go/issues/27415) and [2](https://github.com/golang/go/issues/26650). Considering these discussions are from 2018, seems to confirm that `fsync` guarantees is a quite fascinating topic.

**Comment by jsign** [comment]
Brainstorming mode:
I still think that the `bufio` idea should be thought a little more, since using the default size doesn't seem like a trivial choice. The best size would be one that makes those 5 writes fill in the buffer to later flush the entire thing once. Or maybe make a single slice, put everything in it, and make a single Write (unbuffered). That would mean saving 1 allog from `bufio` and using it in the single slice.

**Comment by prologic** [comment]
> I thought of mocking the `io.Writer` to bound possible latencies there since the idea of this benchmark isn't to test the performance of disk writes, so seems fair. But one gopher gave simpler advice: use `/tmp` to store the database files and mount `/tmp` in RAM (which I did editing `/etc/fstab`).

I actually suspected this all along. And I'm fine with that. Disk writes on systems can be rather unpredictable unless you carefully design a storage system for that purpose; like NVMe + Fast SSDs in some sensible write optimized RAID.

**Comment by prologic** [comment]
> Brainstorming mode:
> I still think that the `bufio` idea should be thought a little more, since using the default size doesn't seem like a trivial choice. The best size would be one that makes those 5 writes fill in the buffer to later flush the entire thing once. Or maybe make a single slice, put everything in it, and make a single Write (unbuffered). That would mean saving 1 alloc from `bufio` and using it in the single slice.

Yeah I think we can try doing this; I have no opinion on which might be better here but my gut tells me pre-allocating a large enough buffer at once and writing all the bits in one single Write() might be faster and reduce allocs?

Also... Why do we bother with putting the Checksum after the Value? Is there a reason for this? Could we just simplify the encoding (_which would break/change the on-disk format again_) to just:

```
KeySize
ValueSize
Checksum
Key
Value
```

**Comment by awnumar** [comment]
> Why do we bother with putting the Checksum after the Value? Is there a reason for this? Could we just simplify the encoding (which would break/change the on-disk format again) to just:

How does that help?

**Comment by prologic** [comment]
Oh it  very likely doesn't help with this particular issue at hand; but I believe it would simplify the code:

https://github.com/prologic/bitcask/blob/7204a33512d91010835e2ad063dafe05e68f1b37/internal/codec.go#L30-L63

My thinking here is restructing the format to have all the binary encoded values first means you can collapse the writes to just 3. I may be wrong :)

**Comment by jsign** [comment]
Yes-ish, but only when the whole encoded Entry is greater than the default buffer of `bufio.Writer` (which is 4096), if isn't the case it would be only 1 write.

In addition to that (minor changes):
* We left an unnecessary `[]byte(msg.Key)`
* also `bufChecksumSize ` can be deleted to reutilize `bufKeyValue` since, again, `checksum` size is lower than `value` size.
* the `bufio.Writer` could be reutilized between instances since it has a `.Reset(io.Writer)` to keep taking advantage of the created inner buffer (and avoid always creating a new one).

**Comment by prologic** [comment]
@jsign Do you have a PR pending to improve this situation?

**Comment by prologic** [comment]
/me just making sure we close out two important issues soon :)

**Comment by jsign** [comment]
> @jsign Do you have a PR pending to improve this situation?

Not really at the moment. Most of my previous bullet-point post was merged [here](https://github.com/prologic/bitcask/pull/64).

**Comment by prologic** [comment]
Can we close this then?

**Comment by jsign** [comment]
Yes, I think so.
